wire/protobuf/serializer.go

```diff
@@ -0,0 +1,182 @@
+	case *client.ChannelSyncMsg:
+		protoEnv.Msg, err = fromChannelSyncMsg(msg)
+	}
+	if err != nil {
+		return err
+	}
+	protoEnv.Sender, protoEnv.Recipient, err = marshalSenderRecipient(env)
+	if err != nil {
+		return err
+	}
+
+	return writeEnvelope(w, protoEnv)
+}
+
+func marshalSenderRecipient(env *wire.Envelope) ([]byte, []byte, error) {
+	sender, err := env.Sender.MarshalBinary()
+	if err != nil {
+		return nil, nil, errors.WithMessage(err, "marshalling sender address")
+	}
+	recipient, err := env.Recipient.MarshalBinary()
+	return sender, recipient, errors.WithMessage(err, "marshalling recipient address")
+}
+
+func writeEnvelope(w io.Writer, env *Envelope) error {
+	data, err := proto.Marshal(env)
+	if err != nil {
+		return errors.Wrap(err, "marshalling envelope")
+	}
+	if err := binary.Write(w, binary.BigEndian, uint16(len(data))); err != nil {
+		return errors.Wrap(err, "writing length to wire")
+	}
+	_, err = w.Write(data)
+	return errors.Wrap(err, "writing data to wire")
+}
+
+// Decode decodes an envelope from the reader, that was encoded using protocol
+// buffers serialization format.
+func (serializer) Decode(r io.Reader) (env *wire.Envelope, err error) { //nolint: funlen, cyclop
+	env = &wire.Envelope{}
+
+	protoEnv, err := readEnvelope(r)
+	if err != nil {
+		return nil, err
+	}
+
+	env.Sender, env.Recipient, err = unmarshalSenderRecipient(protoEnv)
+	if err != nil {
+		return nil, err
+	}
+
+	switch protoMsg := protoEnv.Msg.(type) {
+	case *Envelope_PingMsg:
+		env.Msg = toPingMsg(protoMsg)
+	case *Envelope_PongMsg:
+		env.Msg = toPongMsg(protoMsg)
+	case *Envelope_ShutdownMsg:
+		env.Msg = toShutdownMsg(protoMsg)
+	case *Envelope_AuthResponseMsg:
+		env.Msg = &wire.AuthResponseMsg{}
+	case *Envelope_LedgerChannelProposalMsg:
+		env.Msg, err = toLedgerChannelProposalMsg(protoMsg)
+	case *Envelope_SubChannelProposalMsg:
+		env.Msg, err = toSubChannelProposalMsg(protoMsg)
+	case *Envelope_VirtualChannelProposalMsg:
+		env.Msg, err = toVirtualChannelProposalMsg(protoMsg)
+	case *Envelope_LedgerChannelProposalAccMsg:
+		env.Msg, err = toLedgerChannelProposalAccMsg(protoMsg)
+	case *Envelope_SubChannelProposalAccMsg:
+		env.Msg = toSubChannelProposalAccMsg(protoMsg)
+	case *Envelope_VirtualChannelProposalAccMsg:
+		env.Msg, err = toVirtualChannelProposalAccMsg(protoMsg)
+	case *Envelope_ChannelProposalRejMsg:
+		env.Msg = toChannelProposalRejMsg(protoMsg)
+	case *Envelope_ChannelUpdateMsg:
+		env.Msg, err = toChannelUpdateMsg(protoMsg)
+	case *Envelope_VirtualChannelFundingProposalMsg:
+		env.Msg, err = toVirtualChannelFundingProposalMsg(protoMsg)
+	case *Envelope_VirtualChannelSettlementProposalMsg:
+		env.Msg, err = toVirtualChannelSettlementProposalMsg(protoMsg)
+	case *Envelope_ChannelUpdateAccMsg:
+		env.Msg = toChannelUpdateAccMsg(protoMsg)
+	case *Envelope_ChannelUpdateRejMsg:
+		env.Msg = toChannelUpdateRejMsg(protoMsg)
+	case *Envelope_ChannelSyncMsg:
+		env.Msg, err = toChannelSyncMsg(protoMsg)
```

Comment:
Please throw an error in the default case.

Reply:
Add in this commit: d995e181.